src/script/descriptor.cpp

```diff
@@ -949,14 +955,17 @@ class DescriptorImpl : public Descriptor
 
     bool ToPrivateString(const SigningProvider& arg, std::string& out) const override
     {
-        bool ret = ToStringHelper(&arg, out, StringType::PRIVATE);
+        bool has_priv_key{false};
+        // ToStringHelper should never fail for StringType::PRIVATE,
+        // because it falls back to StringType::PUBLIC when no private key is available.
+        assert(ToStringHelper(&arg, out, StringType::PRIVATE, &has_priv_key));
```

Comment:
In 49f41141c674f83d0b66262a5b21aad623704264 "descriptor: ToPrivateString() pass if  at least 1 priv key exists"

`assert`s should not have side effects. If we are calling a function to actually do something, it should not be in an `assert`.

However, if the return value of `ToStringHelper` is always supposed to be `true`, then it should not be returning anything anyways. Furthermore, I don't see why `ToStringHelper` doesn't return whether the string has a private key, as we are doing in this function. That would allow us to drop `has_priv_key` from all of the function.

Reply:
I refactored the code as suggested and was able to drop `has_priv_key` from the parameter list. Now, `has_priv_key` is returned if `type == StringType::PRIVATE`.